File: Poisk/tools/cas2wav/sources/cas2wav.cpp
// CAS2WAV v1.1 for Poisk PC. by Tronix (C) 2013
// Compile with Virtual Pascal v2.1
// maybe Delphi or Free Pascal

// 2015 Aleksey Morozov, ported to c++

#include "stdafx.h"
#include <stdint.h>
#include <stddef.h>
#include <stdlib.h>
#include <iostream>
#include <io.h>
#include <fcntl.h>
#include <math.h>
#include <vector>
#include <windows.h>

class file_t {
public:
	HANDLE h;

	file_t() { h = INVALID_HANDLE_VALUE; }
	bool open(const wchar_t* name) { close(); h = CreateFile(name, GENERIC_READ,FILE_SHARE_READ|FILE_SHARE_WRITE,0,OPEN_EXISTING,0,0); return h != INVALID_HANDLE_VALUE; }
	bool create(const wchar_t* name) { close(); h = CreateFile(name, GENERIC_READ|GENERIC_WRITE,FILE_SHARE_READ|FILE_SHARE_WRITE,0,CREATE_ALWAYS,0,0); return h != INVALID_HANDLE_VALUE; }
	int read(void* b, size_t l) { DWORD rl; return ReadFile(h, b, l, &rl, 0) ? rl : -1; }
	int write(const void* b, size_t l) { DWORD rl; return WriteFile(h, b, l, &rl, 0) ? rl : -1; }
	int64_t size() { LARGE_INTEGER r; return GetFileSizeEx(h, &r) ? int64_t(r.QuadPart) : int64_t(-1); }
	int64_t seek(int64_t pos, unsigned m) { LARGE_INTEGER s, r; s.QuadPart = pos; return SetFilePointerEx(h, s, &r, m) ? r.QuadPart : -1; }
	void close() { if(h != INVALID_HANDLE_VALUE) { CloseHandle(h); h = INVALID_HANDLE_VALUE; } }
	~file_t() { close(); }
};

const int OUTPUT_FREQUENCY  = 8000;
const int BAUDRATE = 1200;
const int LONG_PULSE_ = 1000; //1
const int SHORT_PULSE_ = 2000; //0
const int LONG_PULSE__ = OUTPUT_FREQUENCY / (BAUDRATE * LONG_PULSE_ / 1200);
const int SHORT_PULSE__ = OUTPUT_FREQUENCY / (BAUDRATE * SHORT_PULSE_ / 1200);

uint16_t Crc16Table[256] = {
    0x0000, 0x1021, 0x2042, 0x3063, 0x4084, 0x50A5, 0x60C6, 0x70E7,
    0x8108, 0x9129, 0xA14A, 0xB16B, 0xC18C, 0xD1AD, 0xE1CE, 0xF1EF,
    0x1231, 0x0210, 0x3273, 0x2252, 0x52B5, 0x4294, 0x72F7, 0x62D6,
    0x9339, 0x8318, 0xB37B, 0xA35A, 0xD3BD, 0xC39C, 0xF3FF, 0xE3DE,
    0x2462, 0x3443, 0x0420, 0x1401, 0x64E6, 0x74C7, 0x44A4, 0x5485,
    0xA56A, 0xB54B, 0x8528, 0x9509, 0xE5EE, 0xF5CF, 0xC5AC, 0xD58D,
    0x3653, 0x2672, 0x1611, 0x0630, 0x76D7, 0x66F6, 0x5695, 0x46B4,
    0xB75B, 0xA77A, 0x9719, 0x8738, 0xF7DF, 0xE7FE, 0xD79D, 0xC7BC,
    0x48C4, 0x58E5, 0x6886, 0x78A7, 0x0840, 0x1861, 0x2802, 0x3823,
    0xC9CC, 0xD9ED, 0xE98E, 0xF9AF, 0x8948, 0x9969, 0xA90A, 0xB92B,
    0x5AF5, 0x4AD4, 0x7AB7, 0x6A96, 0x1A71, 0x0A50, 0x3A33, 0x2A12,
    0xDBFD, 0xCBDC, 0xFBBF, 0xEB9E, 0x9B79, 0x8B58, 0xBB3B, 0xAB1A,
    0x6CA6, 0x7C87, 0x4CE4, 0x5CC5, 0x2C22, 0x3C03, 0x0C60, 0x1C41,
    0xEDAE, 0xFD8F, 0xCDEC, 0xDDCD, 0xAD2A, 0xBD0B, 0x8D68, 0x9D49,
    0x7E97, 0x6EB6, 0x5ED5, 0x4EF4, 0x3E13, 0x2E32, 0x1E51, 0x0E70,
    0xFF9F, 0xEFBE, 0xDFDD, 0xCFFC, 0xBF1B, 0xAF3A, 0x9F59, 0x8F78,
    0x9188, 0x81A9, 0xB1CA, 0xA1EB, 0xD10C, 0xC12D, 0xF14E, 0xE16F,
    0x1080, 0x00A1, 0x30C2, 0x20E3, 0x5004, 0x4025, 0x7046, 0x6067,
    0x83B9, 0x9398, 0xA3FB, 0xB3DA, 0xC33D, 0xD31C, 0xE37F, 0xF35E,
    0x02B1, 0x1290, 0x22F3, 0x32D2, 0x4235, 0x5214, 0x6277, 0x7256,
    0xB5EA, 0xA5CB, 0x95A8, 0x8589, 0xF56E, 0xE54F, 0xD52C, 0xC50D,
    0x34E2, 0x24C3, 0x14A0, 0x0481, 0x7466, 0x6447, 0x5424, 0x4405,
    0xA7DB, 0xB7FA, 0x8799, 0x97B8, 0xE75F, 0xF77E, 0xC71D, 0xD73C,
    0x26D3, 0x36F2, 0x0691, 0x16B0, 0x6657, 0x7676, 0x4615, 0x5634,
    0xD94C, 0xC96D, 0xF90E, 0xE92F, 0x99C8, 0x89E9, 0xB98A, 0xA9AB,
    0x5844, 0x4865, 0x7806, 0x6827, 0x18C0, 0x08E1, 0x3882, 0x28A3,
    0xCB7D, 0xDB5C, 0xEB3F, 0xFB1E, 0x8BF9, 0x9BD8, 0xABBB, 0xBB9A,
    0x4A75, 0x5A54, 0x6A37, 0x7A16, 0x0AF1, 0x1AD0, 0x2AB3, 0x3A92,
    0xFD2E, 0xED0F, 0xDD6C, 0xCD4D, 0xBDAA, 0xAD8B, 0x9DE8, 0x8DC9,
    0x7C26, 0x6C07, 0x5C64, 0x4C45, 0x3CA2, 0x2C83, 0x1CE0, 0x0CC1,
    0xEF1F, 0xFF3E, 0xCF5D, 0xDF7C, 0xAF9B, 0xBFBA, 0x8FD9, 0x9FF8,
    0x6E17, 0x7E36, 0x4E55, 0x5E74, 0x2E93, 0x3EB2, 0x0ED1, 0x1EF0
};

#pragma pack(push, 1)

struct Wave_Fmt {
    uint8_t   RIFFId[4];   // : Array [1..4] of Char;
    uint32_t  RIFFLen;     // : LongInt;
    uint8_t   WAV_Id[4];   // : Array [1..4] of Char;
    uint8_t   Fmt_Id[4];   // : Array [1..4] of Char;
    uint32_t  F_Len;       // : LongInt;
    uint16_t  Format;      // : Word;
    uint16_t  Channels;    // : Word;
    uint32_t  Sample_Rate; // : LongInt;
    uint32_t  Bytes_secnd; // : LongInt;
    uint16_t  Block_Align; // : Word;
    uint16_t  BPS;         // : Word;
    uint8_t   Data_Id[4];  // : Array [1..4] of Char;
    uint32_t  Data_Len;    // : LongInt;
};

struct Poisk_hdr { 
    uint8_t  Magic;        // : Byte; // A5h
    uint8_t  FName[8];     // : Array [1..8] of Char;
    uint8_t  FType;        // : Byte;
    uint16_t FLen;         // : Word;
    uint16_t Seg;          // : Word;
    uint16_t Ofs;          // : Word;
};

#pragma pack(pop)

/*
Var
   F,FF  : File;
   Hdr   : Wave_Fmt;
   Psk   : Poisk_hdr;
   Size  : LongInt;
   s     : String;
   Buf   : Array [1..64000] of Byte;
   BLen  : LongInt;
   Mas   : Array [0..255] of Byte;
   k,kk  : Byte;
*/
uint8_t   buf[32768];
size_t    blen = 0;
file_t    ff;

/*
function IntToHex(IntValue: cardinal; Digits: integer): string;
var
  i, j: cardinal;
  s: string;
begin
  s := '';
  if IntValue > 0 then
  begin
    i := IntValue;
    while i>0 do
    begin
      j := i mod 16;
      if (j<10) then
        s := Chr(Ord('0') + j) + s
      else
        s := Chr(Ord('A') + (j - 10)) + s;
      i := i div 16;
    end;
  end;
  if Digits < 1 then Digits := 1;
  i := Digits - Length(s);
  while i>0 do
  begin
    s := '0' + s;
    i := i - 1;
  end;
  IntToHex := s;
end;

function HexToInt(Str : string): integer;
var i, r : longint;
begin
  val('$'+Str,r, i);
  if i<>0 then HexToInt := 0
  else HexToInt := r;
end;
*/
uint16_t checkCRC(uint8_t* mas) // X^16+X^12+X^5+1
{
   uint16_t crc2 = 0xffff;
   for(int i=0; i<256; i++)
      crc2 = (crc2 << 8) ^ Crc16Table[((crc2 >> 8) ^ mas[i]) & 0xFF];
   return crc2 ^ 0xffff;
}

void writePulse1(int32_t ilen)
{
	if(blen + ilen >= sizeof(buf)) ff.write(buf, blen), blen = 0;
	unsigned ilen2 = ilen / 2;
	memset(buf + blen, 16, ilen2);
	memset(buf + blen + ilen2, -16, ilen - ilen2);
	blen += ilen;
}

void writeHeader(unsigned s)
{
	for(unsigned l = s * (BAUDRATE / 1200); l; l--)
		writePulse1(LONG_PULSE__);
}

void writeByte(uint8_t b)
{
   for(int i=0; i<8; i++, b <<= 1) 
      writePulse1((b & 0x80) ? LONG_PULSE__ : SHORT_PULSE__);
}

void writeBlock(uint8_t* mas)
{
   for(int i=0; i<256; i++) writeByte(mas[i]);
   uint16_t crc = checkCRC(mas);
   writeByte(uint8_t(crc >> 8));
   writeByte(uint8_t(crc));
}

int _tmain(int argc, _TCHAR* argv[])
{
	printf("CAS2WAV v1.1 for Poisk PC. by Tronix (C) 2013\r\n");
	if(argc < 2) 
	{
		printf("Usage:\r\n");
		printf("  CAS2WAV <cas_file> <wav_file> [-n NAME] [-t TYPE] [-s SEG] [-o OFS]\r\n");
        printf("Keys:\r\n");
        printf("  -n NAME    : Set program name on tape (8 chars). Ex.: -n BASIC\r\n");
        printf("  -t TYPE    : Set program type (00,01,40,80,A0).  Ex.: -t 80\r\n");
        printf("  -s SEGMENT : Set memory segment.                 Ex.: -s 0060\r\n");
        printf("  -o OFFSET  : Set memory offset.                  Ex.: -o 081E\r\n");
        printf("Defaults values:\r\n");
        printf("NAME= PROGRAMM, TYPE= 80, SEGMENT= 0060, OFFSET: 081E\r\n");
        return 1;
	}

	file_t f;
	if(!f.open(argv[1]))
	{
		printf("Error open CAS file\r\n");
		return 1;
	}
	std::vector<uint8_t> src;
	size_t src_size = f.size();
	src.resize(src_size);
	if(src_size > 0) 
	{
		if(f.read(&src[0], src_size) != src_size)
		{
			printf("Error open CAS file\r\n");
			return 1;
		}
	}
	f.close();

	if(!ff.create(argv[2]))
	{
		printf("Error create WAV file\r\n");
		return 1;
	}

	Poisk_hdr Psk;
	uint8_t   mas[256];
	Psk.Magic = 0xA5;
	memcpy(Psk.FName, "BASIC\x00\x00", 8);
	Psk.FType = 0x80;
	Psk.FLen  = src.size();
	Psk.Seg   = 0x60;
	Psk.Ofs   = 0x81E;
 /*
   if(argc > 2) 
      For k := 3 to ParamCount do
         Begin
            If ParamStr(k)[1] = '-' then
               Case UpCase(ParamStr(k)[2]) of
                  'N' : begin
                           s := ParamStr(k+1);
                           If Length(s) < 8 then
                              for kk := 1 to 8-Length(s) do s := s + ' ';
                           Move(s[1],Psk.FName,8);
                           WriteLn('Programm name : ',s);
                        end;
                  'T' : begin
                           Psk.FType := HexToInt(ParamStr(k+1));
                           WriteLn('Programm type : ',IntToHex(Psk.FType,2));
                        end;
                  'S' : begin
                           Psk.Seg := HexToInt(ParamStr(k+1));
                           WriteLn('Memory segment: ',IntToHex(Psk.Seg,4));
                        end;
                  'O' : begin
                           Psk.Ofs := HexToInt(ParamStr(k+1));
                           WriteLn('Memory offset : ',IntToHex(Psk.Ofs,4));
                        end;
               Else
                  Error('Invalid key '+ParamStr(k));
               End;
         End;
*/
	printf("Making wav    : %s\n", argv[2]);
	Wave_Fmt hdr;
	memcpy(hdr.RIFFId, "RIFF", 4);
	hdr.RIFFLen = 0;
	memcpy(hdr.WAV_Id, "WAVE", 4);
	memcpy(hdr.Fmt_Id, "fmt ", 4);
	hdr.F_Len = 16;
	hdr.Format = 1; //PCM
	hdr.Channels = 1; // Mono
	hdr.Sample_Rate = OUTPUT_FREQUENCY;
	hdr.Bytes_secnd = OUTPUT_FREQUENCY;
	hdr.Block_Align = 1;
	hdr.BPS = 8;
	memcpy(hdr.Data_Id, "data", 4);
	hdr.Data_Len = 0;

	std::cout << "Format        : " << hdr.Format << std::endl;
	std::cout << "Channels      : " << hdr.Channels << std::endl;
	std::cout << "Sample rate   : " << hdr.Sample_Rate << std::endl;
	std::cout << "Bit per sound : " << hdr.BPS << std::endl;
	ff.write(&hdr, sizeof(hdr));

	blen = 0;

	printf("Write leader1 : ");
	writeHeader(2048);
	writePulse1(SHORT_PULSE__);
	writeByte(0x16);
	std::cout << "Ok!" << std::endl;

	std::cout << "Write fileinfo: ";
	memset(mas, 0, sizeof(mas));
	memcpy(mas, &Psk, 16);
	writeBlock(mas);
	std::cout << "Ok!" << std::endl;

	std::cout << "Close stream  : ";
	for(int k=0; k<4; k++)
		writeByte(0xFF);
	std::cout << "Ok!" << std::endl;

	printf("Write leader2 : ");
	writeHeader(2048);
	writePulse1(SHORT_PULSE__);
	writeByte(0x16);
	std::cout << "Ok!" << std::endl;

	printf("Write data    : ");
	for(unsigned o=0; o<src.size(); o+=256)
	{
		memset(mas, 0, sizeof(mas));
		memcpy(mas, &src[o], o+256 < src.size() ? 256 : src.size()-o);
		writeBlock(mas);
	}
	std::cout << "Ok!" << std::endl;

	if(blen) ff.write(buf, blen);

	// Write size info in WAV header
	uint32_t s = uint32_t(ff.size());
	hdr.RIFFLen = s - 8;
	hdr.Data_Len = s - sizeof(hdr);
	ff.seek(0,0);
	ff.write(&hdr, sizeof(hdr));
 
	ff.close();

	std::cout << std::endl << "Done. Size = " << s << " bytes" << std::endl;
	return 0;
}

File: Poisk/tools/cas2wav/sources/cas2wav.pas
// CAS2WAV v1.1 for Poisk PC. by Tronix (C) 2013
// Compile with Virtual Pascal v2.1
// maybe Delphi or Free Pascal

Const

   OUTPUT_FREQUENCY  = 43200;
   BAUDRATE = 1200;
   LONG_PULSE      = 1000; //1
   SHORT_PULSE       = 2000; //0

Crc16Table: array[0..255] of WORD = (
    $0000, $1021, $2042, $3063, $4084, $50A5, $60C6, $70E7,
    $8108, $9129, $A14A, $B16B, $C18C, $D1AD, $E1CE, $F1EF,
    $1231, $0210, $3273, $2252, $52B5, $4294, $72F7, $62D6,
    $9339, $8318, $B37B, $A35A, $D3BD, $C39C, $F3FF, $E3DE,
    $2462, $3443, $0420, $1401, $64E6, $74C7, $44A4, $5485,
    $A56A, $B54B, $8528, $9509, $E5EE, $F5CF, $C5AC, $D58D,
    $3653, $2672, $1611, $0630, $76D7, $66F6, $5695, $46B4,
    $B75B, $A77A, $9719, $8738, $F7DF, $E7FE, $D79D, $C7BC,
    $48C4, $58E5, $6886, $78A7, $0840, $1861, $2802, $3823,
    $C9CC, $D9ED, $E98E, $F9AF, $8948, $9969, $A90A, $B92B,
    $5AF5, $4AD4, $7AB7, $6A96, $1A71, $0A50, $3A33, $2A12,
    $DBFD, $CBDC, $FBBF, $EB9E, $9B79, $8B58, $BB3B, $AB1A,
    $6CA6, $7C87, $4CE4, $5CC5, $2C22, $3C03, $0C60, $1C41,
    $EDAE, $FD8F, $CDEC, $DDCD, $AD2A, $BD0B, $8D68, $9D49,
    $7E97, $6EB6, $5ED5, $4EF4, $3E13, $2E32, $1E51, $0E70,
    $FF9F, $EFBE, $DFDD, $CFFC, $BF1B, $AF3A, $9F59, $8F78,
    $9188, $81A9, $B1CA, $A1EB, $D10C, $C12D, $F14E, $E16F,
    $1080, $00A1, $30C2, $20E3, $5004, $4025, $7046, $6067,
    $83B9, $9398, $A3FB, $B3DA, $C33D, $D31C, $E37F, $F35E,
    $02B1, $1290, $22F3, $32D2, $4235, $5214, $6277, $7256,
    $B5EA, $A5CB, $95A8, $8589, $F56E, $E54F, $D52C, $C50D,
    $34E2, $24C3, $14A0, $0481, $7466, $6447, $5424, $4405,
    $A7DB, $B7FA, $8799, $97B8, $E75F, $F77E, $C71D, $D73C,
    $26D3, $36F2, $0691, $16B0, $6657, $7676, $4615, $5634,
    $D94C, $C96D, $F90E, $E92F, $99C8, $89E9, $B98A, $A9AB,
    $5844, $4865, $7806, $6827, $18C0, $08E1, $3882, $28A3,
    $CB7D, $DB5C, $EB3F, $FB1E, $8BF9, $9BD8, $ABBB, $BB9A,
    $4A75, $5A54, $6A37, $7A16, $0AF1, $1AD0, $2AB3, $3A92,
    $FD2E, $ED0F, $DD6C, $CD4D, $BDAA, $AD8B, $9DE8, $8DC9,
    $7C26, $6C07, $5C64, $4C45, $3CA2, $2C83, $1CE0, $0CC1,
    $EF1F, $FF3E, $CF5D, $DF7C, $AF9B, $BFBA, $8FD9, $9FF8,
    $6E17, $7E36, $4E55, $5E74, $2E93, $3EB2, $0ED1, $1EF0);

Type
    Wave_Fmt      = Record
      RIFFId      : Array [1..4] of Char;
      RIFFLen     : LongInt;
      WAV_Id      : Array [1..4] of Char;
      Fmt_Id      : Array [1..4] of Char;
      F_Len       : LongInt;
      Format      : Word;
      Channels    : Word;
      Sample_Rate : LongInt;
      Bytes_secnd : LongInt;
      Block_Align : Word;
      BPS         : Word;

      Data_Id     : Array [1..4] of Char;
      Data_Len    : LongInt;
    End;

    Poisk_hdr     = Packed Record
      Magic       : Byte; // A5h
      FName       : Array [1..8] of Char;
      FType       : Byte;
      FLen        : Word;
      Seg         : Word;
      Ofs         : Word;
    End;

Var
   F,FF  : File;
   Hdr   : Wave_Fmt;
   Psk   : Poisk_hdr;
   Size  : LongInt;
   s     : String;
   Buf   : Array [1..64000] of Byte;
   BLen  : LongInt;
   Mas   : Array [0..255] of Byte;
   k,kk  : Byte;

Procedure Error(ss: string);
Begin
   WriteLn(#13#10'Fatal error   : ',ss);
   Halt(255);
End;

function IntToHex(IntValue: cardinal; Digits: integer): string;
var
  i, j: cardinal;
  s: string;
begin
  s := '';
  if IntValue > 0 then
  begin
    i := IntValue;
    while i>0 do
    begin
      j := i mod 16;
      if (j<10) then
        s := Chr(Ord('0') + j) + s
      else
        s := Chr(Ord('A') + (j - 10)) + s;
      i := i div 16;
    end;
  end;
  if Digits < 1 then Digits := 1;
  i := Digits - Length(s);
  while i>0 do
  begin
    s := '0' + s;
    i := i - 1;
  end;
  IntToHex := s;
end;

function HexToInt(Str : string): integer;
var i, r : longint;
begin
  val('$'+Str,r, i);
  if i<>0 then HexToInt := 0
  else HexToInt := r;
end;

Function CheckCRC: Word;  //X^16+X^12+X^5+1
Var
  i : Word;
  crc2 : Word;
Begin
  crc2 := $ffff;
  for I := 0 to 255 do
    crc2 := word(crc2 shl 8) xor Crc16Table[word(crc2 shr 8) xor Mas[i]];
   CheckCRC := crc2 xor $ffff;
End;

Procedure WritePulse(f : longint);
Var
   len, scale : Real;
   n : Longint;
   b : Byte;
Begin
  len := OUTPUT_FREQUENCY/(BAUDRATE*(f/1200));
  scale  := 2.0*PI/len;

  For n := 0 to Round(len)-1 do
      Begin
         b := Byte(Round(sin(n*scale)*127)) xor 128;
         Buf[Blen] := b;
         Inc(Blen);
         If Blen > 64000 then
            begin
               BlockWrite(FF,Buf,64000);
               Blen := 1;
            end;
         Inc(Size);
      End;

End;

Procedure WriteHeader(s : longint);
Var
   i : longint;
Begin
   For i := 0 to Round(s*(BAUDRATE/1200)) do WritePulse(LONG_PULSE);
End;

Procedure WriteByte(b : Byte);
Var
   i : Byte;
Begin
   For i := 7 downto 0 do
      If (b shr i) and 1 = 1 then
         WritePulse(LONG_PULSE)
      else
         WritePulse(SHORT_PULSE);
End;

Procedure WriteBlock;
Var
   i : Byte;
   crc : Word;
Begin
   For i := 0 to 255 do WriteByte(Mas[i]);
   crc := CheckCRC;
   WriteByte(hi(crc));
   WriteByte(lo(crc));
End;

Begin
   WriteLn('CAS2WAV v1.1 for Poisk PC. by Tronix (C) 2013'#13#10);
   If ParamCount < 2 then
      begin
         WriteLn(' Usage:');
         WriteLn('   CAS2WAV <cas_file> <wav_file> [-n NAME] [-t TYPE] [-s SEG] [-o OFS]'#13#10);
         WriteLn(' Keys:');
         WriteLn('   -n NAME    : Set program name on tape (8 chars). Ex.: -n BASIC');
         WriteLn('   -t TYPE    : Set program type (00,01,40,80,A0).  Ex.: -t 80');
         WriteLn('   -s SEGMENT : Set memory segment.                 Ex.: -s 0060');
         WriteLn('   -o OFFSET  : Set memory offset.                  Ex.: -o 081E'#13#10);
         WriteLn(' Defaults values:');
         WriteLn(' NAME= PROGRAMM, TYPE= 80, SEGMENT= 0060, OFFSET: 081E');
         Halt(1);
      end;

   Assign(F,ParamStr(1));
   {$I-}
   Reset(F,1);
   {$I+}
   If IOResult <> 0 then Error('Error open CAS file');

   Assign(FF,ParamStr(2));
   {$I-}
   Rewrite(FF,1);
   {$I+}
   If IOResult <> 0 then Error('Error create WAV file');

   Psk.Magic := $A5;
   Psk.FName := 'PROGRAMM';
   Psk.FType := 128;
   Psk.FLen := FileSize(F);
   Psk.Seg := $0060;
   Psk.Ofs := $081E;

   If ParamCount > 2 then
      For k := 3 to ParamCount do
         Begin
            If ParamStr(k)[1] = '-' then
               Case UpCase(ParamStr(k)[2]) of
                  'N' : begin
                           s := ParamStr(k+1);
                           If Length(s) < 8 then
                              for kk := 1 to 8-Length(s) do s := s + ' ';
                           Move(s[1],Psk.FName,8);
                           WriteLn('Programm name : ',s);
                        end;
                  'T' : begin
                           Psk.FType := HexToInt(ParamStr(k+1));
                           WriteLn('Programm type : ',IntToHex(Psk.FType,2));
                        end;
                  'S' : begin
                           Psk.Seg := HexToInt(ParamStr(k+1));
                           WriteLn('Memory segment: ',IntToHex(Psk.Seg,4));
                        end;
                  'O' : begin
                           Psk.Ofs := HexToInt(ParamStr(k+1));
                           WriteLn('Memory offset : ',IntToHex(Psk.Ofs,4));
                        end;
               Else
                  Error('Invalid key '+ParamStr(k));
               End;
         End;

   WriteLn('Making wav    : ',ParamStr(2));
   Hdr.RIFFId := 'RIFF';
   Hdr.RIFFLen := 0;
   Hdr.WAV_Id := 'WAVE';
   Hdr.Fmt_Id := 'fmt ';
   Hdr.F_Len := 16;
   Hdr.Format := 1; //PCM
   Hdr.Channels := 1; // Mono
   Hdr.Sample_Rate := OUTPUT_FREQUENCY;
   Hdr.Bytes_secnd := OUTPUT_FREQUENCY;
   Hdr.Block_Align := 1;
   Hdr.BPS := 8;
   Hdr.Data_Id := 'data';
   Hdr.Data_Len := 0;

   WriteLn('Format        : ',Hdr.Format);
   WriteLn('Channels      : ',Hdr.Channels);
   WriteLn('Sample rate   : ',Hdr.Sample_Rate);
   WriteLn('Bit per sound : ',Hdr.BPS,#13#10);
   BlockWrite(FF,Hdr,44);

   Size := 1;
   Blen := 1;

   Write(  'Write leader1 : ');
   WriteHeader(2048);
   WritePulse(SHORT_PULSE);
   WriteByte($16);
   WriteLn('Ok!');


   Write(  'Write fileinfo: ');

   FillChar(Mas,256,0);
   Move(Psk,Mas,16);
   WriteBlock;
   WriteLn('Ok!');

   Write(  'Close stream  : ');
   For k := 1 to 4 do WriteByte($FF);
   WriteLn('Ok!'#13#10);

   Write(  'Write leader2 : ');
   WriteHeader(2048);
   WritePulse(SHORT_PULSE);
   WriteByte($16);
   WriteLn('Ok!');

   Write(  'Write data    : ');
   While not Eof(F) do
    begin
      FillChar(Mas,256,0);
      If FilePos(F) < FileSize(F) - 256 then
         BlockRead(F,Mas,256)
      else
         BlockRead(F,Mas,FileSize(F) - FilePos(F));
      WriteBlock;
    end;
   WriteLn('Ok!');

   If Blen <= 64000 then
      BlockWrite(FF,Buf,Blen);

   // Write size info in WAV header
   Seek(FF,0);
   Hdr.RiffLen := Size+44-8;
   Hdr.Data_Len := Size;
   BlockWrite(FF,Hdr,44);

   Close(F);
   Close(FF);
   WriteLn(#13#10'Done. Size = ',size,' bytes');
End.
